File: 0001-0500/0040/0040.combination-sum-ii.js
/*
 * @lc app=leetcode id=40 lang=javascript
 *
 * [40] Combination Sum II
 */

// @lc code=start
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function (candidates, target) {
    let result = [];
    let current = [];
    candidates.sort((a, b) => a - b);
    helper(candidates, target, 0, current, result);
    return result;
}

var helper = function (candidates, target, start, current, result) {
    if (target === 0) {
        result.push(current.slice());
        return;
    }
    for (let i = start; i < candidates.length; i++) {
        if (candidates[i] > target) {
            break;
        }
        if (i > start && candidates[i] === candidates[i - 1]) {
            continue;
        }
        current.push(candidates[i]);
        helper(candidates, target - candidates[i], i + 1, current, result);
        current.pop();
    }
};
// @lc code=end

File: 0001-0500/0040/0040.combination-sum-ii.cs
/*
 * @lc app=leetcode id=40 lang=csharp
 *
 * [40] Combination Sum II
 */

// @lc code=start
public class Solution {
    public IList<IList<int>> CombinationSum2(int[] candidates, int target) {
        var result = new List<IList<int>>();
        Array.Sort(candidates);
        Backtrack(result, new List<int>(), candidates, target, 0);
        return result;
    }

    private void Backtrack(IList<IList<int>> result, List<int> tempList, int[] candidates, int remain, int start) {
        if (remain < 0) return;
        else if (remain == 0) result.Add(new List<int>(tempList));
        else {
            for (int i = start; i < candidates.Length; i++) {
                if (i > start && candidates[i] == candidates[i - 1]) continue;
                tempList.Add(candidates[i]);
                Backtrack(result, tempList, candidates, remain - candidates[i], i + 1);
                tempList.RemoveAt(tempList.Count - 1);
            }
        }
    }
}
// @lc code=end

